Tests/GPU/Locking/main.cpp

```diff
@@ -0,0 +1,119 @@
+#include <AMReX.H>
+#include <AMReX_MultiFab.H>
+#include <AMReX_Gpu.H>
+#include <AMReX_Utility.H>
+#include <AMReX_Array.H>
+#include <AMReX_CudaContainers.H>
+#include <AMReX_ParmParse.H>
+
+using namespace amrex;
+
+void LockingTest();
+
+int main (int argc, char* argv[])
+{
+
+    amrex::Initialize(argc,argv);
+    LockingTest();
+    amrex::Finalize();
+
+}
+
+AMREX_GPU_DEVICE
+void addone(double &num)
+{
+  int blockId = blockIdx.x + blockIdx.y * gridDim.x + gridDim.x * gridDim.y * blockIdx.z;
+  
+  int tid = blockId * (blockDim.x * blockDim.y * blockDim.z)
+    + (threadIdx.z * (blockDim.x * blockDim.y))
+    + (threadIdx.y * blockDim.x) + threadIdx.x ;
+
+  int i = amrex::get_state(tid);
+  num = num+1;
+  amrex::free_state(tid);  
+}
+
+
+void LockingTest ()
+{
+    int Ndraw=1e7;
+
+#ifdef AMREX_USE_CUDA    
+    
+    amrex::Print() << "Generating random numbers using GPU ";
+    amrex::Print() << amrex::Gpu::Device::deviceId() << " on rank ";
+    amrex::Print() << amrex::ParallelDescriptor::MyProc() << "\n";
+
+    Gpu::DeviceVector<double> d_xpos(Ndraw);
+    Gpu::DeviceVector<double> d_ypos(Ndraw);
+    Gpu::DeviceVector<double> d_zpos(Ndraw);
+
+    double *dxpos = d_xpos.dataPtr();
+    double *dypos = d_ypos.dataPtr();
+    double *dzpos = d_zpos.dataPtr();
+    
+
+#endif
+
+    amrex::Vector<double> hx(Ndraw);
+    amrex::Vector<double> hy(Ndraw);
+    amrex::Vector<double> hz(Ndraw);
+    double *hxpos = hx.dataPtr();
+    double *hypos = hy.dataPtr();
+    double *hzpos = hz.dataPtr();
+
+
+    AMREX_PARALLEL_FOR_1D (Ndraw, idx,
+	 {
+#ifdef AMREX_USE_CUDA
+	   
+	   addone(dxpos[idx]);
+	   addone(dypos[idx]);
+	   addone(dzpos[idx]);
```

Comment:
To me it looks like each thread will update it's own value here, regardless of whether the lock is working correctly. If you comment out the get / set state above, does the test still pass? It seems that what you should do is pass a pointer into the nums into the `addone` function, and then increment the value at index `i`. If the total is correct then, we'll know that the lock is preventing race conditions.